probe_src/performance_test.py

```diff
@@ -0,0 +1,81 @@
+import psutil
+import subprocess
+from dataclasses import dataclass
+import datetime
+
+@dataclass
+class Result:
+    returncode: int
+    cpu_times: tuple
+    memory_info: tuple
+    io_counters: tuple
+    stdout: str
+    stderr: str
+    start_time: datetime.datetime
+    end_time: datetime.datetime
+
+def benchmark_command(command: str, warmup_iterations: int, benchmark_iterations: int) -> list[Result]:
+    results = []
+
+    for _ in range(warmup_iterations):
+        subprocess.run(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+    for _ in range(benchmark_iterations):
+        start_time_psutil = datetime.datetime.now()
+
+        proc = psutil.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        cpu_times = memory_info = io_counters = None
+
+        if proc.is_running():
+            try:
+                cpu_times = proc.cpu_times()
+                memory_info = proc.memory_info()
+                io_counters = proc.io_counters()
+            except psutil.NoSuchProcess:
+                break
+            except psutil.AccessDenied:
+                pass
+
+        stdout, stderr = proc.communicate()
+        
+        end_time_psutil = datetime.datetime.now()
+
+        start_time_wait = datetime.datetime.now()
+        returncode = proc.wait()
+        end_time_wait = datetime.datetime.now()
+
+        result = Result(
+            returncode=returncode,
+            cpu_times=cpu_times,
+            memory_info=memory_info,
+            io_counters=io_counters,
+            stdout=stdout.decode('utf-8'),
+            stderr=stderr.decode('utf-8'),
+            start_time=start_time_psutil,
+            end_time=end_time_wait  
+        )
+        results.append(result)
+
+    return results
+
+if __name__ == "__main__":
+    # command and other parameters 
+    command_to_run = "./PROBE record --make head ../flake.nix"
+    warmup_count = 5
```

Comment:
Can we separate this into `record --no-transcribe` and `transcribe` and structure the code such that we can try 3 or 5 different commands (a list).

Reply:
Yes, I will do this, though I think we can't add the `transcribe` command in the list since it requires the path to the raw file.

Reply:
Added more commands in the list, reduced the warmup and benchmark iterations just to get results earlier

Reply:
I'm a bit confused by "we can't add `transcribe` ...". Will something like this work?

```
for each test_case:
    time "PROBE record --no-transcribe test_case"
    output times to CSV
    time "PROBE transcribe test_case"
    output times to CSV
```

Reply:
That is because `transcribe` will require a path to the raw data which it needs to transcribe. The `--no-transcribe` command stores this raw data, and the `transcribe` command later processes it. Each transcription refers to the raw data directory that was created during the recording phase.
So, the `--no-transcribe` command after execution also tells us the path to the raw directory, and we give that path to the `transcribe` command to process the raw data.


`./PROBE record` does both execution and transcription
`./PROBE record --make --no-transcribe` , only executes
`./PROBE transcribe-only /path/to/raw_probe_dir --output probe_log` does only transcription

So, i meant to say that it might not be possible to hard code a transcribe command in the commands list since we don't know the path to the raw directory.

The code you shared can work but we will have to store the path to the raw directory somewhere after each --no-transcribe command.
```
for each test_case:
    time "PROBE record --no-transcribe test_case"
    output times to CSV
    time "PROBE transcribe test_case"
    output times to CSV
```

Reply:
> The code you shared can work but we will have to store the path to the raw directory somewhere after each --no-transcribe command.

Let's go ahead and do that, if it's simple. That will allow us to have better performance timing because we can separate the overhead of recording from transcribing.